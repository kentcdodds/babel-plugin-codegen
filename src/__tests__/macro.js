import path from 'path'
import pluginTester from 'babel-plugin-tester'
import plugin from 'babel-plugin-macros'

const projectRoot = path.join(__dirname, '../../')

expect.addSnapshotSerializer({
  print(val) {
    return val.split(projectRoot).join('<PROJECT_ROOT>/')
  },
  test(val) {
    return typeof val === 'string'
  },
})

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: {filename: __filename, parserOpts: {plugins: ['jsx']}},
  tests: {
    'as tag': `
      import codegen from '../macro'
      const greeting = 'Hello world!'
      codegen\`module.exports = "module.exports = '\${greeting}';"\`
    `,
    'as function': `
      const myCodgen = require('../macro')
      myCodgen(\`
        module.exports = "var x = {booyah() { return 'booyah!'; } };"
      \`)
    `,
    'as jsx': `
      const Codegen = require('../macro')
      const ui = (
        <Codegen>{"module.exports = '<div>Hi</div>'"}</Codegen>
      )
    `,
    'as jsx with tag': `
      const Codegen = require('../macro')
      const ui = (
        <Codegen>{\`module.exports = '<div>Hi</div>'\`}</Codegen>
      )
    `,
    'with multiple': `
      import codegen from '../macro'

      codegen\`module.exports = ['a', 'b', 'c'].map(l => 'export const ' + l + ' = ' + JSON.stringify(l)).join(';')\`
    `,
    'as require call': `
      import codegen from '../macro';
      var x = codegen.require('./fixtures/return-one');
    `,
    'invalid usage: as fn argument': {
      code: `
        import codegen from '../macro';
        var x = doSomething(codegen);
      `,
      error: true,
    },
    'invalid usage: missing code string': {
      code: `
        import codegen from '../macro';
        var x = codegen;
      `,
      error: true,
    },
  },
})
