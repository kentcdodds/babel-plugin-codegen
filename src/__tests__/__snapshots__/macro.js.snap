// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros as function: as function 1`] = `

const myCodgen = require('../macro')
myCodgen(\`
  module.exports = "var x = {booyah() { return 'booyah!'; } };"
\`)

      ↓ ↓ ↓ ↓ ↓ ↓

var x = {
  booyah() {
    return 'booyah!'
  },
}


`;

exports[`macros as jsx with tag: as jsx with tag 1`] = `

const Codegen = require('../macro')
const ui = (
  <Codegen>{\`module.exports = '<div>Hi</div>'\`}</Codegen>
)

      ↓ ↓ ↓ ↓ ↓ ↓

const ui = <div>Hi</div>


`;

exports[`macros as jsx: as jsx 1`] = `

const Codegen = require('../macro')
const ui = (
  <Codegen>{"module.exports = '<div>Hi</div>'"}</Codegen>
)

      ↓ ↓ ↓ ↓ ↓ ↓

const ui = <div>Hi</div>


`;

exports[`macros as require call: as require call 1`] = `

import codegen from '../macro';
var x = codegen.require('./fixtures/return-one');

      ↓ ↓ ↓ ↓ ↓ ↓

var x = 1


`;

exports[`macros as tag: as tag 1`] = `

import codegen from '../macro'
const greeting = 'Hello world!'
codegen\`module.exports = "module.exports = '\${greeting}';"\`

      ↓ ↓ ↓ ↓ ↓ ↓

const greeting = 'Hello world!'
module.exports = 'Hello world!'


`;

exports[`macros invalid usage: as fn argument: invalid usage: as fn argument 1`] = `

import codegen from '../macro';
var x = doSomething(codegen);

      ↓ ↓ ↓ ↓ ↓ ↓

Error: <PROJECT_ROOT>/src/__tests__/macro.js: ../macro: codegen macro must be used as a tagged template literal, function, jsx, or .require call
  1 | import codegen from '../macro';
> 2 | var x = doSomething(codegen);
    |                     ^^^^^^^

`;

exports[`macros invalid usage: missing code string: invalid usage: missing code string 1`] = `

import codegen from '../macro';
var x = codegen;

      ↓ ↓ ↓ ↓ ↓ ↓

Error: <PROJECT_ROOT>/src/__tests__/macro.js: ../macro: codegen macro must be used as a tagged template literal, function, jsx, or .require call
  1 | import codegen from '../macro';
> 2 | var x = codegen;
    |         ^^^^^^^

`;

exports[`macros with a single export declaration: with a single export declaration 1`] = `

import codegen from '../macro'

codegen\`module.exports = "export const a = 'a'"\`

      ↓ ↓ ↓ ↓ ↓ ↓

export const a = 'a'


`;

exports[`macros with a single import declaration: with a single import declaration 1`] = `

import codegen from '../macro'

codegen\`module.exports = "import a from 'a'"\`

      ↓ ↓ ↓ ↓ ↓ ↓

import a from 'a'


`;

exports[`macros with multiple: with multiple 1`] = `

import codegen from '../macro'

codegen\`module.exports = ['a', 'b', 'c'].map(l => 'export const ' + l + ' = ' + JSON.stringify(l)).join(';')\`

      ↓ ↓ ↓ ↓ ↓ ↓

export const a = 'a'
export const b = 'b'
export const c = 'c'


`;
